contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,92 @@
+# Local PV Distributed Static Provisioner
+Authors: @verult, @msau42
+        
+## Design Goals
+1. Limit all PV object interactions to a single master pod in order to minimize node access to the Kubernetes system.
+1. Assume disk deletion is asynchronous, as required by block volumes.
+
+## Related Works
+* Original provisioner design, [here](https://github.com/kubernetes-incubator/external-storage/tree/master/local-volume/provisioner#design).
+* Disk cleaner (deleter) implementation for block volumes, [here](https://github.com/kubernetes-incubator/external-storage/pull/312).
+
+## Design 1: Master-Heavy
+
+### High Level Overview
+The static provisioner will be split into a single master (ideally placed on the master node) and many minions (one per node). The master contains the bulk of provisioner logic, including syncing PV objects with the API server and triggering discovery and deletion. Minions are left to perform the physical discovery and deletion of the local volumes on its node. gRPC is used to communicate between master and minions, with the master as the client and each minion as a separate server.
+
+The master keeps track of `MinionInfo`, a collection containing one entry for each minion keeping track of relevant information. The master also runs a group of goroutines, one for each minion, which controls and interacts with the minions.
+
+### Minion Provisioner
+The minion provisioner exposes the following RPCs:
+
+| RPC                           | Description |
+| --- | --- |
+| `GetMinionName()`             | Returns the minion provisioner's name. |
+| `GetMinionStatus()`           | Returns the minion status, as described in the next section. |
```

Comment:
What about `ListAndWatch()` that would return `stream of VolumeEvent` that returns all existing directories + waits for event for every created / deleted directory?

Using golang syntax instead of protobuf:

```go
type VolumeEvent struct {
    Volume Volume
    EventType string
}

type Volume struct {
    DirectoryName string
    StorageClassName string
    Capacity ...
    ...
}

type EventType string
const EventTypeAdded = "added" /* new directory found */
const EventTypeAdded = "deleted" /* something deleted existing directory, master should mark it as unusable somehow */
```

Reply:
Replying to myself, ListAndWatch needs to be smarter. Master starts, reads existing PVs from API server, runs ListAndWatch and needs to mark all directories that were deleted while the master was offline so user knows that the PV is unusable. It won't get "deleted" event, directory was removed before the master started -> its PV is still marked as available -> troubles.

This is implementation detail though, there are many ways out of this. I still recommend something like `ListAndWatch`.

Reply:
In the master's initial sync with the minion, it can use the lack of a `Volume` to indicate deletion.